[RateLimiter] Charge read to rate limiter in BackupEngine

hx235 opened the pull request:
Context:
While all the non-trivial write operations in BackupEngine go through the RateLimiter, reads currently do not. In general, this is not a huge issue because (especially since some I/O efficiency fixes) reads in BackupEngine are mostly limited by corresponding writes, for both backup and restore. But in principle we should charge the RateLimiter for reads as well.

Summary
- Charged read operations in `BackupEngineImpl::CopyOrCreateFile`, `BackupEngineImpl::ReadFileAndComputeChecksum`, `BackupEngineImpl::BackupMeta::LoadFromFile` and `BackupEngineImpl::GetFileDbIdentities`

Test:
- Passed existing tests
- Passed added unit tests

zhichao-cao commented on utilities/backupable/backupable_db_test.cc:
```diff
@@ -2620,6 +2621,102 @@ TEST_P(BackupEngineRateLimitingTestWithParam, RateLimiting) {
+  std::int64_t total_bytes_through_with_no_read_charge =
+      restore_rate_limiter->GetTotalBytesThrough();
+  CloseBackupEngine();
+  AssertBackupConsistency(0, 0, 10, 20);
```
Yes.

hx235 replied:
Thanks!

pdillinger commented on utilities/backupable/backupable_db.cc:
```diff
@@ -928,13 +928,15 @@ BackupEngineImpl::BackupEngineImpl(const BackupEngineOptions& options,
-        NewGenericRateLimiter(options_.backup_rate_limit));
+    options_.backup_rate_limiter.reset(NewGenericRateLimiter(
+        options_.backup_rate_limit, 100 * 1000 /* refill_period_us */,
+        10 /* fairness */, RateLimiter::Mode::kAllIo /* mode */));
```
I think this is an unnecessarily disruptive change in behavior, because there's an inconsistency in default behavior with NewGenericRateLimiter. Specifically, NewGenericRateLimiter only limits writes by default, whereas here we are changing the default for backup+restore to limit both reads and writes.

I would prefer either:
(a - easy and effective) Leave existing options as they are but comment on `BackupEngineOptions::backup_rate_limit` that the limit only applies to writes, and you must set `backup_rate_limiter` to have an option to also limit reads.
(b) Add option `bool backup_rate_limit_reads` to enable limiting reads under the limit given by `backup_rate_limit`

hx235 replied:
Thanks for the suggestion - I will go with (a) for both backup and restore

pdillinger commented on utilities/backupable/backupable_db.cc:
```diff
@@ -928,13 +928,15 @@ BackupEngineImpl::BackupEngineImpl(const BackupEngineOptions& options,
-        NewGenericRateLimiter(options_.restore_rate_limit));
+    options_.restore_rate_limiter.reset(NewGenericRateLimiter(
+        options_.restore_rate_limit, 100 * 1000 /* refill_period_us */,
+        10 /* fairness */, RateLimiter::Mode::kAllIo /* mode */));
```
Same for restore limit

pdillinger reviewed (changes requested):
Also need to apply to BackupEngineImpl::ReadFileAndComputeChecksum, which is the most important case where there's no corresponding write.

There's also BackupEngineImpl::BackupMeta::LoadFromFile and BackupEngineImpl::GetFileDbIdentities. Overall these are small, but could be noticeable from an instantaneous rate perspective.

I haven't looked at the tests in detail (I'm not too worried about perfection as this is not data correctness critical), but you should probably also add test for rate limiting `VerifyBackup(verify_with_checksum=true)`

pdillinger commented:
Also update HISTORY.md. I would argue this is a "bug fix" because the old behavior would violate natural expectation.

hx235 commented:
> Also update HISTORY.md. I would argue this is a "bug fix" because the old behavior would violate natural expectation.

Thanks for the reminder! Will do!

hx235 commented on utilities/backupable/backupable_db.cc:
```diff
@@ -550,7 +551,8 @@ class BackupEngineImpl {
                              const std::string& file_path, std::string* db_id,
-                             std::string* db_session_id);
+                             std::string* db_session_id,
+                             RateLimiter* rate_limiter);
```
[Question]
I saw in your [PR](https://github.com/facebook/rocksdb/pull/8732) that backup_rate_limiter is assumed to be used in ReadFileAndComputeChecksum and I am not sure if you'd prefer me doing the same things for GetFileDbIdentities and LoadFromFile, assuming they are going to use backup_rate_limiter only.

I called LoadFromFile with backup_rate_limiter in BackupEngineImpl::Initialize() just because it looks to me that it is in the context of dealing with backup files - not sure if it is a general fact that LoadFromFile will only be used with backup_rate_limiter.

hx235 commented:
Update:
- Charged read in `BackupEngineImpl::ReadFileAndComputeChecksum`, `BackupEngineImpl::BackupMeta::LoadFromFile` and `BackupEngineImpl::GetFileDbIdentities` as requested
- Added unit-test on rate limiting `VerifyBackup(verify_with_checksum=true` as suggested and others to cover above changes
- Reverted changes to default 'GenericRateLimiter' created when no rate limiter is passed in, as suggested
- Updated HISTORY.md as suggested
- Should be ready for review after signals are clear

pdillinger commented on utilities/backupable/backupable_db.cc:
```diff
@@ -2619,6 +2633,10 @@ Status BackupEngineImpl::BackupMeta::LoadFromFile(
   // Failures handled at the end
   std::string line;
   if (backup_meta_reader->ReadLine(&line)) {
+    if (rate_limiter != nullptr) {
```
Seems like it would be easier/cleaner to add rate limiting support to LineFileReader.

hx235 replied:
Yes - added this as a motivation in the task to add support for rate limiting in file system wrapper layer or file system layer.

pdillinger commented on utilities/backupable/backupable_db_test.cc:
```diff
@@ -2618,6 +2620,176 @@ TEST_P(BackupEngineRateLimitingTestWithParam, RateLimiting) {
   AssertBackupConsistency(0, 0, 100000, 100010);
 }
+
+TEST_P(BackupEngineRateLimitingTestWithParam, RateLimitingVerifyBackup) {
```
Probably historical to avoid slower tests in valgrind to save time. I really wish people would put in 'why' comments also.

pdillinger commented on utilities/backupable/backupable_db_test.cc:
```diff
@@ -2618,6 +2620,176 @@ TEST_P(BackupEngineRateLimitingTestWithParam, RateLimiting) {
+  if (makeThrottler) {
+    EXPECT_GE(verify_backup_time, 0.8 * rate_limited_verify_backup_time);
+  } else {
+    EXPECT_LT(verify_backup_time, rate_limited_verify_backup_time);
```
Unfortunately, we cannot have tests that check execution time is below some threshold, because this will be flaky. At least remove that part.

These rate limiting tests should really be upgraded to use something other than elapsed clock time for verification, especially since these new tests will likely add quite a bit of running time to the backup tests. They could use either GetTotalRequests() or mock time (time_elapse_only_sleep etc.) as in RateLimiterTest.AutoTuneIncreaseWhenFull. One challenge is when specifying just a rate, the limiter object is hidden from view, but it should be possible to make it accessible to the running test.

But improving the rate limiter test speed and accuracy could be a follow-up PR.

hx235 replied:
> One challenge is when specifying just a rate, the limiter object is hidden from view, but it should be possible to make it accessible to the running test.

Yes - that was the main challenge I had when implementing this test and I guess we can do something like sync point to make it accessible.

> These rate limiting tests should really be upgraded to use something other than elapsed clock time for verification, especially since these new tests will likely add quite a bit of running time to the backup tests. 

Opened a task for this.

hx235 replied:
Offline discussion: we will need something to check the existing behavior of VerifyBackup is not affected and this can be included in the follow-up improvement PR mentioned above.

pdillinger reviewed (changes requested):
* One fix required to avoid flaky test.
* Fix required for HISTORY.md.
* Other suggestions optional or for follow-up PR.

hx235 commented on HISTORY.md:
```diff
@@ -7,6 +7,8 @@
 * Fix a race in DumpStats() with column family destruction due to not taking a Ref on each entry while iterating the ColumnFamilySet.
 * Fix a race in item ref counting in LRUCache when promoting an item from the SecondaryCache.
 * Fix a race in BackupEngine if RateLimiter is reconfigured during concurrent Restore operations.
+* Fix a bug that backup_rate_limit and restore_rate_limit in BackupEngine only apply to writes but not including reads.
```
Good thoughts! My description is more about the case where user expects to limit both read and write by passing a rate limiter that can do that but finds out that only write is limited. But now reading your suggestion makes me realize yours is a more precise description for the bug since (1) in the scenario I described, read is somehow also limited by the symmetric property of read and write in Backup engine (2) yours is a more direct and easy-to-understand scenario. 

Will update that according to your suggestion - thanks! 👍

hx235 commented on utilities/backupable/backupable_db.cc:
```diff
@@ -407,7 +407,8 @@ class BackupEngineImpl {
         Logger* info_log,
-        std::unordered_set<std::string>* reported_ignored_fields);
+        std::unordered_set<std::string>* reported_ignored_fields,
+        RateLimiter* rate_limiter);
```
Thanks for pointing this out - absolutely helpful for a C++ newbie like me!

hx235 commented on utilities/backupable/backupable_db.cc:
```diff
@@ -2274,6 +2284,10 @@ Status BackupEngineImpl::GetFileDbIdentities(Env* src_env,
     } else {
       table_properties = tp.get();
+      if (table_properties != nullptr && rate_limiter != nullptr) {
+        rate_limiter->Request(sizeof(*table_properties), Env::IO_LOW,
```
Sure - will do!

hx235 commented:
Update
- Removed flaky test check and included that as part of the rate limiting test improvement in BackupEngine
- Reworded HISTORY.md description
- Adopted suggestions on C++ coding style and clarification comment
- Should be ready for review after the signals are clear

pdillinger reviewed (approved):
👍 👍

hx235 commented:
Update:
- Resolved HISTORY.md conflict on Github UI

hx235 commented:
Update:
- Manually solve conflict on HISTORY.md
